addons/mrp/models/stock_move.py

```diff
@@ -291,6 +291,18 @@ def create(self, vals_list):
         return super().create(vals_list)
 
     def write(self, vals):
+        for move in self:
+            is_product_id_updated = 'product_id' in vals and move.product_id.id != vals.get('product_id')
+            is_move_valid_and_updated = is_product_id_updated and move.production_id and move.state not in ['draft', 'cancel', 'done']
+
+            if is_move_valid_and_updated:
+                moves_data = move.copy_data()
+                for move_data in moves_data:
+                    move_data.update({'product_id': vals.get('product_id')})
+                updated_product_move = move.create(moves_data)
+                move.unlink()
+                updated_product_move._action_confirm()
+                self = updated_product_move
```

Comment:
You update `self` inside a loop on `self` which can be problematic.
Also introducing a loop on self makes the `move.create()`, `move.unlink()` and `update_product_move._action_confirm()` much less efficient. The previous diff was better, only a filter on `move in self` to only manage moves with a different product than the one in `vals` was needed.

Reply:
Hello @Whenrow

Reply:
Hello @Whenrow 
Have you checked the last changes?
Thanks!